openingd/: Fail `fundchannel_start` if we already are, or will become, the fundee.

ZmnSCPxj opened the pull request:
Fixes: #4108

Requesting review from nifty and rusty, since my grasp of the openingd is not good.

As well, there may be a similar issue in dual-funding, so please consider as well for that point-of-view, I am much more ignorant about the dual-funding flow than openingd flow.

niftynei commented on lightningd/opening_common.c:
```diff
@@ -53,6 +53,8 @@ new_uncommitted_channel(struct peer *peer)
 	uc->peer->uncommitted_channel = uc;
 	tal_add_destructor(uc, destroy_uncommitted_channel);
 
+	uc->is_fundee = false;
```
I'd suggest calling this `got_offer`

niftynei commented on openingd/openingd.c:
```diff
@@ -974,6 +974,14 @@ static u8 *fundee_channel(struct state *state, const u8 *open_channel_msg)
+	 * `fundchannel_start` on this peer, and fail all future ones
+	 * as well.
+	 */
+	msg = towire_openingd_will_be_fundee(NULL);
```
You don't need a new wire for this, see `opening_got_offer` in `lightningd/opening_control.c`.

niftynei commented on lightningd/opening_control.c:
```diff
@@ -1086,6 +1121,10 @@ static struct command_result *json_fund_channel_start(struct command *cmd,
 	}
 
+	if (peer->uncommitted_channel->is_fundee) {
+		return command_fail(cmd, LIGHTNINGD, "We are the fundee");
```
```suggestion
		return command_fail(cmd, LIGHTNINGD, "Have in-progress `open_channel` from peer");
```

niftynei reviewed (changes requested):
Nice debugging work on this! You don't need a new wire call to fix this. We already do a similar/exact call a few lines above, `opening_got_offer`, which is used for the `openchannel` hook. We should toggle this flag there.

ZmnSCPxj commented:
The reason why I did not reuse `got_offer` is because of this:

https://github.com/ElementsProject/lightning/blob/60be62a3bbc25dc849430baaaed4d61c1b4720b5/openingd/openingd.c#L965-L972

We would need to add more code in `lightningd` to clear the `is_fundee`/`got_offer` flag in case we end up rejecting it, I think (not sure, `opening` flow is unfamiliar to me).

Looking over at `lightningd/opening_control.c`, it seems to me I have to set the `got_offer` flag *after* this check, so if we got into this part of the code, we reject before we set the `got_offer` flag:

https://github.com/ElementsProject/lightning/blob/60be62a3bbc25dc849430baaaed4d61c1b4720b5/lightningd/opening_control.c#L729-L736

Then later, if a plugin tells us to fail it, we should add a check in this function:

https://github.com/ElementsProject/lightning/blob/60be62a3bbc25dc849430baaaed4d61c1b4720b5/lightningd/opening_control.c#L625-L644

Unfortunately, it seems we do not have access to the `struct uncommitted_channel` here to clear the `got_offer` flag, so I should modify `openchannel_hook_payload` to include that, is that correct?

ZmnSCPxj commented on lightningd/opening_control.c:
```diff
@@ -415,6 +420,17 @@ static void opening_funder_finished(struct subd *openingd, const u8 *resp,
 }
 
+static void
+opening_funder_failed_cancel_commands(struct uncommitted_channel *uc,
```
I think `opening_fundee_started` is just fine where it is, and the other one is also just fine where it is.  A small declaration to me is just small and easy to skip over with proper folding.  *shrug*.

ZmnSCPxj replied:
Also, modern vim understands etags just fine and we have a `make TAGS` that uses etags, declarations do not confuse etags the way they confuse ctags.

rustyrussell replied:
@niftynei is right here; C is always written bottom to top.  Sometimes judicious use of pre-declaration can reduce diff size, but best practice is usually a separate preparatory commit to reorder.

ZmnSCPxj commented:
Removed the new wire.  Cancel a conflicting `fundchannel_start` after plugins all agree everything is kosher.  Renamed `is_fundee` to `got_offer`.  Did ***not*** move function just to avoid pre-declaration, I think avoiding pre-declaration at all costs leads to ***awful*** code layout where later-executed functions are read first before earlier-executed functions, I think the minor cost of a few lines for pre-declaration is far cheaper than the mental gymnastics needed to understand code written in a non-straightforward order.  If I had moved `opening_funder_failed_cancel_commands` above `opening_funder_failed` just to avoid a pre-declaration, then it would have obscured that the new function is just the latter part of the original single function with no modifications.

niftynei commented:
So, not to get too "putting on my arm-chair psychologist hat" here, but as you did bring up the topic of "mental gymnastics" I think we should discuss the laundry list of rationales you've given so far as to why it's ok to disregard the request to do the reorg needed to remove the `static` function declaration.

Rationales offered: 
- "opening_fundee_started is just fine where it is"
- "the other one is also just fine where it is"
- "A small declaration to me is just small and easy to skip over"
- "ctags is really just a lousy beta-level software"
- "Twisting your codebase because of lousy tools is worse than just not having that tool."
- "I think avoiding pre-declaration at all costs leads to awful code layout"
- "the minor cost of a few lines for pre-declaration is far cheaper than the mental gymnastics needed"

These are all stated in the pull request comments. However, I find it peculiar that none of these are mentioned in the actual code comment you added (and vice versa, that the reason given there isn't mentioned elsewhere), which is

```
/* These two functions used to be a single function, pre-declaring
 * here in order to reduce review load since the original function
 * is not really changed at all in its operation, just need access
 * to the cleanup part in a separate piece of code.
 */
```

Which can be summarized as
- "reduce review load"

I'm not going to address the rationales given here, I merely wish to point out that you've triggered my "methinksth thou dost protesth too much" heuristic of argumentation. Usually this is an indication that there's actually another, unspoken and perhaps unconscious, rationale for as to why you'd like to keep the code the way it is, rather than following the fairly small request, I think at this point, to change it.

What do you think?

ZmnSCPxj commented:
Well, I could put a long treatise in the comments like so, if that is what you would prefer:

```c
/* Often, in programming, we find we need to split functions into smaller pieces,
 * simply as a consequence of refactoring code for code reuse, or in a server
 * context, simply as a consequence of having to do callback-style coding
 * in order to block for a task but not block the entire server.
 * Now, ideally you want those functions to be in a reasonable order, that is,
 * in the same order you would have read the lines if they were a single
 * large function instead of broken up into tiny pieces.
 *
 * However, C requires functions to be predeclared before first use.
 * So if you have to split up a function into two pieces, for refactoring, or
 * to make a callback for a blocking operation, and the first piece has to
 * refer to the second piece (to call it or to provide as callback function)
 *
 * - Either you predeclare the second piece and add a few lines.
 * - You twist your code so that the second piece is written before the first.
 *
 * Now, there are other reasons to predeclare, or not predeclare.
 * 
 * - If you have to do mutual recursion, no choice but to predeclare.
 * - If you want to isolate a common operation off into its own function,
 *   you should not predeclare, or should consider moving it off into
 *   its own source file.
 *   (And if it is off into its own source file, it is effectively predeclared
 *   in its own header).
 *
 * On the other hand, maintaining the proper order for executing
 * code should generally be a reason to predeclare code.
 * Predeclaration is just a few lines.
 *
 * But really, the issue is that C requires predeclaration *at all*.
 */
```

Does that satisfy your psychologist?  Or can we get on with fixing this bug?

ZmnSCPxj commented:
Removed the predeclaration as a separate commit so that the code-move can be reviewed separately from the code-modification.

niftynei commented on lightningd/opening_control.c:
```diff
@@ -595,6 +569,27 @@ opening_funder_failed_cancel_commands(struct uncommitted_channel *uc,
 	uc->fc = tal_free(uc->fc);
 }
+static void
+opening_funder_failed_cancel_commands(struct uncommitted_channel *uc,
```
I think you meant to remove this?

niftynei commented on tests/test_connection.py:
```diff
@@ -2695,3 +2695,19 @@ def test_htlc_retransmit_order(node_factory, executor):
+
+    l1.rpc.connect(l2.info['id'], 'localhost', l2.port)
+
+    l1.rpc.fundchannel_start(l2.info['id'], 100000)
```
if you call 'fundchannel' here, instead of `fundchannel_start`, then at some point in the future (soonTM), it'll also test the v2 channel open code.

ZmnSCPxj replied:
It has to be `fundchannel_start`, since the race condition being tested is when both peers perform `fundchannel_start`.  Otherwise once `fundchannel` returns, the state is different and a different non-buggy part of the code is triggered.

Adding the changes below results in the test passing even without code changes, so the test *has to* use `fundchannel_start`:

```diff
     '''
     l1, l2 = node_factory.get_nodes(2)
+    l1.fundwallet(10**6 + 1000000)
 
     l1.rpc.connect(l2.info['id'], 'localhost', l2.port)
 
-    l1.rpc.fundchannel_start(l2.info['id'], 100000)
+    l1.rpc.fundchannel(l2.info['id'], 100000)
 
     fut = executor.submit(l2.rpc.fundchannel_start, l1.info['id'], 100000)
     with pytest.raises(RpcError):
```

niftynei commented:
bump @ZmnSCPxj

niftynei reviewed (approved):
ACK 1ace2ad

Rebased on master to resolve conflict with (merged) #4126
